src/D2L.Security.OAuth2/Keys/Default/Data/JwksProvider.cs

```diff
@@ -1,65 +1,118 @@
 			}
 		}
 
-		string IJwksProvider.Namespace {
-			get {
-				return m_jwksEndpoint.AbsoluteUri;
+		async Task<JsonWebKey> IJwksProvider.RequestJwkAsync( Guid keyId ) {
+			var url = GetJwkEndpoint( m_authEndpoint, keyId );
+			try {
+				using( var res = await m_httpClient.GetAsync( url ) ) {
+					// This is temporary while we try to fully deprecate the
+					// JWKS route. 404 might mean the key doesn't exist (which
+					// will make the call to jwks likely return 200 but still
+					// result in no key - that's fine but slow) or the jwk
+					// route isn't supported which will make this recover (but
+					// slowely.) Where it matters (the LMS and auth) JWK is
+					// supported so this shouldn't be necessary. We don't
+					// expect many "legitimate" 404s (keys that don't exist.)
+					// so in practice this should only happen when it's
+					// actually important, if it ever happens.
+					if( res.StatusCode == HttpStatusCode.NotFound ) {
+						return await FallbackToJwks( keyId );
+					}
+
+					res.EnsureSuccessStatusCode();
+
+					string json = await res.Content
+						.ReadAsStringAsync()
+						.SafeAsync();
+
+					return JsonWebKey.FromJson( json);
+				}
+			} catch( HttpRequestException e ) {
+				throw CreateException( e, url );
+			} catch( JsonWebKeyParseException e ) {
+				throw CreateException( e, url );
 			}
 		}
 
-		private Exception CreateException( Exception e ) {
-			string message = string.Format(
-				"Error while looking up JWKS at {0}: {1}",
-				m_jwksEndpoint,
-				e.Message
-			);
+		private async Task<JsonWebKey> FallbackToJwks( Guid keyId ) {
+			var allKeys = await ( this as IJwksProvider ).RequestJwksAsync();
+
+			JsonWebKey theKey;
+
+			if( !allKeys.TryGetKey( keyId, out theKey ) ) {
+				throw new PublicKeyLookupFailureException(
+					$"couldn't find public key {keyId} after falling back to jwks route"
+				);
+			}
+
+			return theKey;
+		}
+
+		string IJwksProvider.Namespace => m_authEndpoint.AbsoluteUri;
+
+		private Exception CreateException( Exception e, Uri endpoint ) {
+			string message = $"Error while looking up key(s) at {endpoint}: {e.Message}";
 
 			return new PublicKeyLookupFailureException( message, e );
 		}
 
-		private static Uri BuildJwksEndpoint( Uri authEndpoint ) {
-			string authRoot = authEndpoint.ToString();
-			if( !authRoot.EndsWith( "/" ) ) {
-				authRoot += "/";
-			}
+		private static Uri GetJwkEndpoint( Uri authEndpoint, Guid keyId ) {
+			string authRoot = MakeSureThereIsATrailingSlash( authEndpoint );
 
-			authRoot += JWKS_PATH;
+			authRoot += $"jwk/{keyId}";
 
 			return new Uri( authRoot );
 		}
+
+		private static Uri GetJwksEndpoint( Uri authEndpoint ) {
+			string authRoot = MakeSureThereIsATrailingSlash( authEndpoint );
+
+			authRoot += ".well-known/jwks";
+
+			return new Uri( authRoot );
+		}
+
+		private static string MakeSureThereIsATrailingSlash( Uri uri ) {
+			string root = uri.ToString();
+			if ( root[root.Length - 1 ] == '/') {

```

Comment:
Looking through the source code it could happen in irrelevant cases but not when there is a host (the only case we'll care about.)
I think I'm ok with an NRE in that case... good catch though.